File: 128-longest-consecutive-sequence/longest-consecutive-sequence.rb
# @param {Integer[]} nums
# @return {Integer}
def longest_consecutive(nums)
    # inputs: unsorted array of integers
    # outputs: length of longest consecutive element

    # example: [1,5,3,2] = 3
    #          [] = 0
    #          [2,4,6] = 1
    #          [-1,-2,-3] = 3
    
    longest = 0
    nums = nums.to_set


    nums.each do |num|
        if !nums.include?(num - 1)
            current_num = num
            current_length = 1
            while nums.include?(current_num + 1)
                current_num += 1
                current_length += 1
            end

            longest = [longest, current_length].max
        end
    end

    longest
end

File: 128-longest-consecutive-sequence/longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # we are given an UNSORTED array of ints
        # RETURN: want the length of the longest consecutive element sequence, which will be an int
        # what is considered a valid sequence? if they are increasing by 1
        # cant sort since that is O(nlogn), how do we get an O(n) algorithm
        # duplicate numbers do not count as part of the longest sequence

        # Max sequeunce can ever only be len(nums)

        visited = set(nums)

        longest_seq = 0

        if len(nums) == 0:
            return 0

        for num in visited:
            if (num - 1) not in visited:
                current = num
                count = 1
                while (current + 1) in visited:
                    count += 1
                    current += 1
                
                longest_seq = max(longest_seq, count)


        return longest_seq